Strata/Transform/LoopElim.lean

```diff
@@ -0,0 +1,68 @@
+/-
+  Copyright Strata Contributors
+
+  SPDX-License-Identifier: Apache-2.0 OR MIT
+-/
+
+import Strata.DL.Lambda.LExpr
+import Strata.Languages.Boogie.Statement
+
+namespace Boogie
+open Imperative Lambda
+
+/-! ## Loop elimination
+
+This transformation recursively removes loops from a Boogie statement, resulting
+in a new, acyclic statement. If a loop invariant is present, it adds checks that
+the invariant is established on entry and re-established at the end of each
+iteration.
+
+-/
+
+partial
+def Statement.removeLoopsM (s : Boogie.Statement) : StateM Nat Boogie.Statement :=
```

Comment:
Note: this could be more general, using `Imperative.Stmt` instead of `Boogie.Statement`, but it'll require introduction of a new type class and a few other changes. Because of that, I'm inclined to do it in a later refactoring PR.